cilium/CFP-35631-support-cni-status-operation.md

```diff
@@ -0,0 +1,61 @@
+
+## **Motivation**
+
+General motivation for the introduction of CNI STATUS operation can be found here: [Proposal: "STATUS" verb](https://github.com/containernetworking/cni/issues/859)
+
+This CFP proposes that cilium-cni should add support for the STATUS verb which provides a more detailed way to pass the health of cilium-cni and cilium-agent back to container runtime.
+
+Below is how kubelet determines if a Node is “Network Ready”
+
+* The  CNI configuration file defines how we set up the network for a pod. When a container runtime needs to set up networking for a container, it looks at this file and exec’s the CNI plugin binaries in the config.  
+* The [NetworkReady](https://github.com/kubernetes/kubernetes/blob/34abd36e88e50d913af9413f9353515deb7f09d4/pkg/kubelet/kubelet.go#L2873-L2876) condition in kubelet is used to determine if the pods can be scheduled on the node and is initialized to false when the node starts. Kubelet calls the [CRI Status() API](https://github.com/kubernetes/kubernetes/blob/3718932d9a3ea459217ca11efe1ae9ead6aa92c1/pkg/kubelet/kubelet.go#L2871) periodically and sets this flag to true when the container runtime returns the NetworkReady status. Kubelet continues to use the same API to monitor the node readiness.  
+* The container runtime uses the presence of a CNI configuration file as an indicator of the CNI plugin's readiness to process ADD requests. This inferred status is then reported to kubelet.
+
+Checking for a CNI config file is an unreliable indicator of network health on the node, as various issues can prevent proper network functionality despite the file's existence. Some potential issues include
+
+* cilium-agent cannot be scheduled during upgrade (due to resource constraints on the node).  
+* cilium-agent is crash looping.  
+* CNI config file is invalid.  
+* We run out of IP addresses on the node.
+
+There is a parallel effort in progress in go-cni ([https://github.com/containerd/go-cni/pull/121](https://github.com/containerd/go-cni/pull/121)) and CRI to call CNI STATUS and pass the status to kubelet. This would enable kubelet to set NetworkReady to false and prevent workload pods from being added (KEP is WIP).
+
+## **Goals**
+
+* Implement STATUS in Cilium with the minimal return code 50\.  [“50: The plugin is not available (i.e. cannot service ADD requests)”](https://github.com/containernetworking/CNI/blob/main/SPEC.md#status-check-plugin-status)
+
+## **Non-Goals**
+
+* Define “limited connectivity” and implement return code 51\.  [“51: The plugin is not available, and existing containers in the network may have limited connectivity.”](https://github.com/containernetworking/CNI/blob/main/SPEC.md#status-check-plugin-status)  
```

Comment:
I'm looking for this to be settled before merging the CFP, just to make sure the non-goal documented here is in line with the standard. Otherwise the CFP indeed seems to be "Implementable".

Reply:
Agreed, error code 51 is optional, and it's fine to implement this never, or later.